File: src/main/java/uk/co/thinkofdeath/patchtools/wrappers/ClassSet.java
/*
 * Copyright 2014 Matthew Collins
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package uk.co.thinkofdeath.patchtools.wrappers;

import com.google.common.io.ByteStreams;
import com.google.common.io.Closeables;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.commons.Remapper;
import org.objectweb.asm.commons.RemappingClassAdapter;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodNode;
import uk.co.thinkofdeath.patchtools.PatchScope;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.util.*;

public class ClassSet implements Iterable<String> {

    private final Map<String, ClassWrapper> classes = new HashMap<>();
    private final ClassPathWrapper classPath;

    private boolean simplified;

    public ClassSet(ClassPathWrapper wrapper) {
        classPath = wrapper;
    }

    public void simplify() {
        if (simplified) return;
        simplified = true;
        // Safety copy
        new ArrayList<>(classes.values()).stream()
            .filter(v -> !v.isHidden())
            .forEach(v -> v.getMethods().stream()
                .filter(m -> !m.isHidden())
                .forEach(m -> {
                    MethodNode node = v.getMethodNode(m);
                    if (((node.access & Opcodes.ACC_PUBLIC) != 0
                        || (node.access & Opcodes.ACC_PROTECTED) != 0)
                        && (node.access & Opcodes.ACC_STATIC) == 0) {
                        for (String inter : v.getNode().interfaces) {
                            replaceMethod(m, inter);
                        }
                        replaceMethod(m, v.getNode().superName);
                    }
                }));
    }

    private void replaceMethod(MethodWrapper methodWrapper, String clazz) {
        if (clazz == null) return;
        ClassWrapper cl = getClassWrapper(clazz);
        if (cl == null) {
            return;
        }
        final ClassWrapper finalCl = cl;
        MethodWrapper target = cl.getMethods().stream()
            .filter(m -> m.getName().equals(methodWrapper.getName())
                && m.getDesc().equals(methodWrapper.getDesc()))
            .filter(m -> {
                MethodNode node = finalCl.getMethodNode(m);
                return (((node.access & Opcodes.ACC_PUBLIC) != 0
                    || (node.access & Opcodes.ACC_PROTECTED) != 0)
                    && (node.access & Opcodes.ACC_STATIC) == 0);
            })
            .findFirst().orElse(null);
        if (target != null) {
            if (target.isHidden()) {
                methodWrapper.hidden = true;
            }
            cl.getMethods().remove(target);
            cl.getMethods().add(methodWrapper);
            methodWrapper.add(cl);
        }
        for (String inter : cl.getNode().interfaces) {
            replaceMethod(methodWrapper, inter);
        }
        replaceMethod(methodWrapper, cl.getNode().superName);

    }

    public void add(InputStream clazz) {
        add(clazz, true);
    }

    public void add(InputStream clazz, boolean close) {
        try {
            add(ByteStreams.toByteArray(clazz));
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        } finally {
            if (close) {
                Closeables.closeQuietly(clazz);
            }
        }
    }

    public void add(byte[] clazz) {
        ClassReader classReader = new ClassReader(clazz);
        ClassNode node = new ClassNode(Opcodes.ASM5);
        classReader.accept(node, 0);
        add(node);
    }

    public void add(ClassNode node) {
        classes.put(node.name, new ClassWrapper(this, node));
    }

    public void remove(String name) {
        classes.remove(name);
    }

    public byte[] getClass(String name) {
        ClassSetWriter classWriter = new ClassSetWriter(ClassWriter.COMPUTE_MAXS | ClassWriter.COMPUTE_FRAMES);
        ClassWrapper wrapper = classes.get(name);
        if (wrapper == null || wrapper.isHidden()) {
            return null;
        }
        wrapper.getNode().version = Opcodes.V1_8;
        wrapper.getNode().accept(classWriter);
        return classWriter.toByteArray();
    }

    public byte[] getClass(String name, PatchScope scope) {
        ClassReader classReader = new ClassReader(getClass(name));
        ClassSetWriter classWriter = new ClassSetWriter(ClassWriter.COMPUTE_MAXS | ClassWriter.COMPUTE_FRAMES);
        ClassWrapper wrapper = classes.get(name);
        if (wrapper == null || wrapper.isHidden()) {
            return null;
        }
        wrapper.getNode().version = Opcodes.V1_8;
        classReader.accept(new RemappingClassAdapter(classWriter, new ClassRemapper(scope)), ClassReader.EXPAND_FRAMES);
        return classWriter.toByteArray();
    }

    public ClassWrapper getClassWrapper(String name) {
        ClassWrapper cl = classes.get(name);
        if (cl == null) {
            cl = classPath.find(this, name);
            if (cl == null) return null;
            classes.put(cl.getNode().name, cl);
        }
        return cl;
    }

    public String[] classes() {
        return classes.keySet().toArray(new String[classes.keySet().size()]);
    }

    private String[] hiddenStrippedCache;

    public String[] classes(boolean stripHidden) {
        if (!stripHidden) {
            return classes();
        }
        if (hiddenStrippedCache == null) {
            hiddenStrippedCache = classes.entrySet().stream()
                .filter(v -> !v.getValue().isHidden())
                .map(Map.Entry::getKey)
                .toArray(String[]::new);
        }
        return hiddenStrippedCache;
    }

    @Override
    public Iterator<String> iterator() {
        return classes.keySet().iterator();
    }

    private class ClassSetWriter extends ClassWriter {

        public ClassSetWriter(int flags) {
            super(flags);
        }

        @Override
        protected String getCommonSuperClass(String type1, String type2) {
            if (type1.equals(type2)) {
                return type1;
            }

            HashSet<String> supers = new HashSet<>();
            supers.add(type1);
            supers.add(type2);

            boolean run1 = true;
            boolean run2 = true;
            ClassWrapper t1 = getClassWrapper(type1);
            ClassWrapper t2 = getClassWrapper(type2);

            if (t1 == null || t2 == null) {
                return "java/lang/Object";
            }

            if ((t1.getNode().access & Opcodes.ACC_INTERFACE) != 0
                || (t2.getNode().access & Opcodes.ACC_INTERFACE) != 0) {
                return "java/lang/Object";
            }

            while (run1 || run2) {
                if (run1) {
                    if (t1 == null || t1.getNode().superName == null) {
                        run1 = false;
                    } else {
                        t1 = getClassWrapper(t1.getNode().superName);
                        if (t1 != null && !supers.add(t1.getNode().name)) {
                            return t1.getNode().name;
                        }
                    }
                }
                if (run2) {
                    if (t2 == null || t2.getNode().superName == null) {
                        run2 = false;
                    } else {
                        t2 = getClassWrapper(t2.getNode().superName);
                        if (t2 != null && !supers.add(t2.getNode().name)) {
                            return t2.getNode().name;
                        }
                    }
                }
            }
            return "java/lang/Object";
        }
    }

    private class ClassRemapper extends Remapper {
        private final PatchScope scope;

        public ClassRemapper(PatchScope scope) {
            this.scope = scope;
        }

        @Override
        public String map(String typeName) {
            ClassWrapper cls = getClassWrapper(typeName);
            if (cls != null) {
                String name = scope.getClass(cls);
                return name == null ? typeName : name;
            }
            return typeName;
        }

        @Override
        public String mapMethodName(String owner, String name, String desc) {
            ClassWrapper cls = getClassWrapper(owner);
            if (cls != null) {
                MethodWrapper methodWrapper = cls.getMethod(name, desc);
                String nName = scope.getMethod(methodWrapper);
                return nName == null ? name : nName.split("\\(")[0];
            }
            return name;
        }

        @Override
        public String mapFieldName(String owner, String name, String desc) {
            ClassWrapper cls = getClassWrapper(owner);
            if (cls != null) {
                FieldWrapper fieldWrapper = cls.getField(name, desc);
                String nName = scope.getField(fieldWrapper);
                return nName == null ? name : nName.split("::")[0];
            }
            return name;
        }
    }
}

File: src/main/kotlin/uk/co/thinkofdeath/patchtools/wrappers/ClassSet.kt
/*
 * Copyright 2014 Matthew Collins
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package uk.co.thinkofdeath.patchtools.wrappers

import java.io.InputStream
import org.objectweb.asm.tree.ClassNode
import uk.co.thinkofdeath.patchtools.PatchScope
import org.objectweb.asm.ClassWriter
import org.objectweb.asm.commons.Remapper
import java.util.HashMap
import java.util.ArrayList
import org.objectweb.asm.ClassReader
import org.objectweb.asm.Opcodes
import org.objectweb.asm.commons.RemappingClassAdapter
import java.util.HashSet

public class ClassSet(private val classPath: ClassPathWrapper) : Iterable<String> {

    private val classes = HashMap<String, ClassWrapper>()

    private var simplified: Boolean = false

    public fun simplify() {
        if (simplified) return
        simplified = true
        // Safety copy
        ArrayList(classes.values())
            .filter { !it.isHidden() }
            .forEach {
                val v = it
                it.methods
                    .filter { !it.isHidden() }
                    .forEach {
                        val node = v.getMethodNode(it)!!
                        if (((node.access and Opcodes.ACC_PUBLIC) != 0
                            || (node.access and Opcodes.ACC_PROTECTED) != 0)
                            && (node.access and Opcodes.ACC_STATIC) == 0) {
                            for (inter in v.node.interfaces) {
                                replaceMethod(it, inter)
                            }
                            replaceMethod(it, v.node.superName)
                        }
                    }
            }

        // Second pass to add everything to each other
        ArrayList(classes.values())
            .filter { !it.isHidden() }
            .forEach { grab(it, it) }
    }

    private fun grab(root: ClassWrapper, current: ClassWrapper?) {
        if (current == null) return
        current.fields
            .filter { !it.isHidden() }
            .filter {
                val node = current.getFieldNode(it)!!
                ((node.access and Opcodes.ACC_PUBLIC) != 0
                    || (node.access and Opcodes.ACC_PROTECTED) != 0)
                    && (node.access and Opcodes.ACC_STATIC) == 0
            }
            .forEach { it.add(root) }
        current.methods
            .filter { !it.isHidden() }
            .filter {
                val node = current.getMethodNode(it)!!
                ((node.access and Opcodes.ACC_PUBLIC) != 0
                    || (node.access and Opcodes.ACC_PROTECTED) != 0)
                    && (node.access and Opcodes.ACC_STATIC) == 0
            }
            .forEach { it.add(root) }

        for (inter in current.node.interfaces) {
            grab(root, getClassWrapper(inter))
        }
        grab(root, getClassWrapper(current.node.superName))
    }

    private fun replaceMethod(methodWrapper: MethodWrapper, clazz: String?) {
        if (clazz == null) return
        val cl = getClassWrapper(clazz)
        if (cl == null) {
            return
        }
        val target = cl.methods
            .filter { it.name == methodWrapper.name }
            .filter { it.desc == methodWrapper.desc }
            .filter {
                val node = cl.getMethodNode(it)!!
                (((node.access and Opcodes.ACC_PUBLIC) != 0
                    || (node.access and Opcodes.ACC_PROTECTED) != 0))
                    && (node.access and Opcodes.ACC_STATIC) == 0
            }
            .first
        if (target != null) {
            if (target.isHidden()) {
                methodWrapper.hidden = true
            }
            cl.methods.remove(target)
            cl.methods.add(methodWrapper)
            methodWrapper.add(target)
        }
        for (inter in cl.node.interfaces) {
            replaceMethod(methodWrapper, inter)
        }
        replaceMethod(methodWrapper, cl.node.superName)

    }

    public fun add(clazz: InputStream) {
        add(clazz, true)
    }

    public fun add(clazz: InputStream, close: Boolean) {
        try {
            add(clazz.readBytes())
        } finally {
            if (close) {
                clazz.close()
            }
        }
    }

    public fun add(clazz: ByteArray) {
        val classReader = ClassReader(clazz)
        val node = ClassNode(Opcodes.ASM5)
        classReader.accept(node, 0)
        add(node)
    }

    public fun add(node: ClassNode) {
        classes.put(node.name, ClassWrapper(this, node))
    }

    public fun remove(name: String) {
        classes.remove(name)
    }

    public fun getClass(name: String): ByteArray? {
        val classWriter = ClassSetWriter(ClassWriter.COMPUTE_MAXS or ClassWriter.COMPUTE_FRAMES)
        val wrapper = classes.get(name)
        if (wrapper == null || wrapper.isHidden()) {
            return null
        }
        wrapper.node.version = Opcodes.V1_8
        wrapper.node.accept(classWriter)
        return classWriter.toByteArray()
    }

    public fun getClass(name: String, scope: PatchScope): ByteArray? {
        val classReader = ClassReader(getClass(name))
        val classWriter = ClassSetWriter(0)
        val wrapper = classes.get(name)
        if (wrapper == null || wrapper.isHidden()) {
            return null
        }
        classReader.accept(RemappingClassAdapter(classWriter, ClassRemapper(scope)), ClassReader.EXPAND_FRAMES)
        return classWriter.toByteArray()
    }

    public fun getClassWrapper(name: String?): ClassWrapper? {
        if (name == null) return null
        var cl: ClassWrapper? = classes.get(name)
        if (cl == null) {
            cl = classPath.find(this, name)
            if (cl == null) return null
            classes.put(cl!!.node.name, cl)
        }
        return cl
    }

    public fun classes(): Array<String> {
        return classes.keySet().copyToArray()
    }

    private var hiddenStrippedCache: Array<String>? = null

    public fun classes(stripHidden: Boolean): Array<String> {
        if (!stripHidden) {
            return classes()
        }
        if (hiddenStrippedCache == null) {
            hiddenStrippedCache = classes
                .filterValues { !it.isHidden() }
                .map { it.getKey() }
                .copyToArray()
        }
        return hiddenStrippedCache!!
    }

    override fun iterator(): Iterator<String> {
        return classes.keySet().iterator()
    }

    private inner class ClassSetWriter(flags: Int) : ClassWriter(flags) {

        override fun getCommonSuperClass(type1: String, type2: String): String {
            if (type1 == type2) {
                return type1
            }

            val supers = HashSet<String>()
            supers.add(type1)
            supers.add(type2)

            var run1 = true
            var run2 = true
            var t1 = getClassWrapper(type1)
            var t2 = getClassWrapper(type2)

            if (t1 == null || t2 == null) {
                return "java/lang/Object"
            }

            if ((t1!!.node.access and Opcodes.ACC_INTERFACE) != 0
                || (t2!!.node.access and Opcodes.ACC_INTERFACE) != 0) {
                return "java/lang/Object"
            }

            while (run1 || run2) {
                if (run1) {
                    if (t1 == null || t1!!.node.superName == null) {
                        run1 = false
                    } else {
                        t1 = getClassWrapper(t1!!.node.superName)
                        if (t1 != null && !supers.add(t1!!.node.name)) {
                            return t1!!.node.name
                        }
                    }
                }
                if (run2) {
                    if (t2 == null || t2!!.node.superName == null) {
                        run2 = false
                    } else {
                        t2 = getClassWrapper(t2!!.node.superName)
                        if (t2 != null && !supers.add(t2!!.node.name)) {
                            return t2!!.node.name
                        }
                    }
                }
            }
            return "java/lang/Object"
        }
    }

    private inner class ClassRemapper(private val scope: PatchScope) : Remapper() {

        override fun map(typeName: String): String {
            val cls = getClassWrapper(typeName)
            if (cls != null) {
                return scope.getClass(cls) ?: typeName
            }
            return typeName
        }

        override fun mapMethodName(owner: String, name: String, desc: String): String {
            val cls = getClassWrapper(owner)
            if (cls != null) {
                val methodWrapper = cls.getMethod(name, desc)
                if (methodWrapper != null) {
                    val nName = scope.getMethod(methodWrapper)
                    return if (nName == null) name else nName.split("\\(")[0]
                }
            }
            return name
        }

        override fun mapFieldName(owner: String, name: String, desc: String): String {
            val cls = getClassWrapper(owner)
            if (cls != null) {
                val fieldWrapper = cls.getField(name, desc)
                if (fieldWrapper != null) {
                    val nName = scope.getField(fieldWrapper)
                    return if (nName == null) name else nName.split("::")[0]
                }
            }
            return name
        }
    }
}